### PR TITLE
Add option to print dispersy statistics when a test fails

### PR DESCRIPTION
**Pull request by LipuFei**

### CODE REVIEW

File: Tribler/Test/test_as_server.py [hunk 1]
```diff
@@ -212,11 +212,33 @@ def _shutdown_session(self, session):
 
         self._logger.debug("Session has shut down")
 
-    def assert_(self, boolean, reason=None, do_assert=True):
+    def assert_(self, boolean, reason=None, do_assert=True, tribler_session=None, dump_statistics=False):
```
Comment (NielsZeilemaker) [inline Tribler/Test/test_as_server.py]: You could pass a reference to the tribler_session.dump_statistics method instead

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
@whirm After I made this, the `channel_test` doesn't fail anymore...

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by whirm** [comment]
@LipuFei ready?

**Comment by LipuFei** [comment]
@whirm The print code I added have never been tested. Maybe we should wait?

**Comment by whirm** [comment]
sure

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by NielsZeilemaker** [comment]
@LipuFei something crazy is happening with the titles of the screenshots
http://jenkins.tribler.org//job/GH_Tribler_pull-request-tester_devel/3750/artifact/output//Screenshot-13.png
Is that because of your change?

**Comment by LipuFei** [comment]
@NielsZeilemaker It's another one #1324. I will try to simplify the title

**Comment by NielsZeilemaker** [comment]
Maybe add a newline, and increase the height of the textbox.

2015-03-18 16:46 GMT+01:00 Lipu Fei <notifications@github.com>:

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
retest this please

**Comment by LipuFei** [comment]
@NielsZeilemaker @whirm Here is the statistics of a failed `channel_search` test. We have 14 candidates in the AllChannel community and we have sent out 8 `channelsearch` requests, but we never get any response. I guess that's why it failed.

```
--- 8164f55c2f828738fa779570e4605a81fec95c9d: AllChannelCommunity:
----- Sync bloom created: 7
----- Statistics:
------- delay:
--------- delay_message:DelayMessageByMissingMessage: 1211
--------- delay_packet:Missing member: 806
--------- delay_message:DelayMessageBySequence: 26
------- outgoing:
--------- dispersy-missing-identity: 565
--------- dispersy-missing-sequence: 17
--------- dispersy-missing-message: 812
--------- -caused by missing-identity-: 10
--------- channelsearch: 8
--------- dispersy-introduction-response: 2
--------- dispersy-puncture-request: 2
--------- dispersy-puncture: 4
--------- channelcast-request: 12
--------- dispersy-introduction-request: 11
------- drop:
--------- drop_message:duplicate message by binary packet [dispersy-undo-own]: 8
--------- drop_message:duplicate message by member^global_time (1) [dispersy-undo-own]: 30
--------- delay_timeout:DelayMessageByMissingMessage: 467
------- success:
--------- votecast: 413
--------- dispersy-missing-identity: 10
--------- dispersy-puncture-request: 4
--------- dispersy-introduction-response: 11
--------- dispersy-identity: 564
--------- channelcast: 4
--------- dispersy-puncture: 10
--------- dispersy-undo-own: 656
--------- dispersy-introduction-request: 2
------- created:
--------- dispersy-identity: 1
----- Sync bloom reused: 4
----- Packets Delayed Sent: 1394 / 3712 ~37.6%
----- Member: a249ed654fa733d2e6b82c247e269478fae465d5
----- Candidates: 15 
----- Median global time: 18512730722996 (18512730719646 difference)
----- Packets Delayed Timeout: 467 / 2043 ~22.9%
----- Acceptable range: 10000
----- Packets Delayed Success: 1255 / 2043 ~61.4%
----- Packets Success: 1674 / 3712 ~45.1%
----- Packets Sent: 1443 / 5155 ~28.0%
----- Database id: None
----- Sync bloom skipped: 0
----- Database: ? packets
----- Packets Delayed Received: 2043 / 3712 ~55.0%
----- Global time: 3350
----- Packets Dropped: 505 / 3712 ~13.6%
----- Identifier: 8164f55c2f828738fa779570e4605a81fec95c9d
----- Packets Received: 3712 / 5155 ~72.0%
----- Classification: AllChannelCommunity
----- Packets Created: 1
```

**Comment by NielsZeilemaker** [comment]
It only sends a reply if it has any matching results, so that doesn't actually say anything.
Guys, let's first fix the unit-tests in #1338 the family filter is broken so that needs priority.

**Comment by LipuFei** [comment]
@NielsZeilemaker ok

**Comment by NielsZeilemaker** [comment]
The strange thing is, its receiving responses but its not showing in the statistics:
```
DEBUG   1426847285.73 UserCallbackHandler:92    ucb: notify called: signal_allchannel signal_on_search_results None ({'keywords': [u'mp3'], 'torrents': {'^Y\xf2\xff\x10o\xea\x93\xe4\xab\xf4\\\x8b\x9b\r\x13&\x9d\xc0\xf6': set(['<5\x95\xebl#\xc6x\xcfl\xbd\xd1o\x92\xe1\xc7j{w-']), '\xf3\xeec\x01\x8f\x08\x93_#\x87A\xe1W\x88F\xa2\xc6q\x18\xba': set(['\xa9\xb1wz\x86\xc0\x0eO\x9f\xa68\xd0\x1c%\xea\xf2j\xda\xa4|', "\t\x81\x8eY\xe7\xb0%j'\x92\xe9\x12\xf7\xef\x05\x93\xa9\xcb\xd4\x97", '\xee\x9a\xdf\x95\x18\xc4\xbf\x92\x9b-*\x00?\xc955C\xea\x92%', '\xa8\xd9\xf8hU.\xc5\x06\x04yCbhY\n\x81\x13\xc4\xc4\xee', '\xba\x15\x05iW\x97w\xcdq>\xb9\x013{x\xf3\x13s\xb3\xe5', 'S\x11N\xa2\xe6\x11\x85\x9c\xa3\x06\xca\xec\xb2\x89^*t\xc5\x0e\xd2', '\xb9U\x89\xe6\xdb~\x13q\xb2\xba\xdd\x81\xcb\xde\x13\x05\xad\x0cp\xad'])}},)

**Comment by LipuFei** [comment]
@NielsZeilemaker Good point, I will investigate the `gitHits()` function and the database part.
```
got search response for 'mp3'
UserCallbackHandler:92    ucb: notify called: signal_allchannel signal_on_search_results None ({'keywords': [u'mp3'], 'torrents': {'^Y\xf2\xff\x10o\xea\x93\xe4\xab\xf4\\\x8b\x9b\r\x13&\x9d\xc0\xf6': set(['<5\x95\xebl#\xc6x\xcfl\xbd\xd1o\x92\xe1\xc7j{w-']), '\xf3\xeec\x01\x8f\x08\x93_#\x87A\xe1W\x88F\xa2\xc6q\x18\xba': set(['\xa9\xb1wz\x86\xc0\x0eO\x9f\xa68\xd0\x1c%\xea\xf2j\xda\xa4|', "\t\x81\x8eY\xe7\xb0%j'\x92\xe9\x12\xf7\xef\x05\x93\xa9\xcb\xd4\x97", '\xee\x9a\xdf\x95\x18\xc4\xbf\x92\x9b-*\x00?\xc955C\xea\x92%', '\xa8\xd9\xf8hU.\xc5\x06\x04yCbhY\n\x81\x13\xc4\xc4\xee', '\xba\x15\x05iW\x97w\xcdq>\xb9\x013{x\xf3\x13s\xb3\xe5', 'S\x11N\xa2\xe6\x11\x85\x9c\xa3\x06\xca\xec\xb2\x89^*t\xc5\x0e\xd2', '\xb9U\x89\xe6\xdb~\x13q\xb2\xba\xdd\x81\xcb\xde\x13\x05\xad\x0cp\xad'])}},)
   GuiDBHandler:145   GUIDBHandler: Task(TopSearchPanel.py:184 (ShowSearching)) took to be called 0.3 (expected 0.2), actual task took 0.0 guiTaskQueue
      community:2009  got 1 incoming packets
       database:218   disabling commit [/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/Test/_test_.Tribler/sqlite/dispersy.db]
   GuiDBHandler:111   GUIDBHandler: adding Task(FakeResult)
 TimedTaskQueue:103   ttqueue: EVENT IN QUEUE 1426847286.39 <function SearchGridManager.py:250 (getHitsInCategory) at 0x7f466ec0ee60>
 TimedTaskQueue:107   ttqueue: EVENT NOT TILL 0.65774679184
 TimedTaskQueue:56    ttqueue: ADD EVENT 0.25 <function TopSearchPanel.py:193 (FakeResult) at 0x7f466eb7e578> TopSearchPanel.py:193 (FakeResult)
       database:242   enabling commit without committing now [/home/jenkins/workspace/GH_Tribler_pull-request-tester_devel/tribler/Tribler/Test/_test_.Tribler/sqlite/dispersy.db]
      community:1932  delay a 543 byte packet/message (DelayMessageReqChannelMessage) from {192.168.11.3:7759 59.17.121.239:7759}
 TimedTaskQueue:103   ttqueue: EVENT IN QUEUE 1426847285.99 <function TopSearchPanel.py:193 (FakeResult) at 0x7f466eb7e578>
 TimedTaskQueue:107   ttqueue: EVENT NOT TILL 0.248702049255
   GuiDBHandler:111   GUIDBHandler: adding Task(RemoteSearchManager_refresh_channel_[u'mp3'])
SearchGridManager:1739  ChannelManager: searchLocalDB: returning old hit list 0
SearchGridManager:1667  ChannelManager: getChannelHits: search found: 0 items
SearchGridManager:1710  ChannelManager: getChannelHits took 0.000257015228271
```

**Comment by LipuFei** [comment]
@NielsZeilemaker The statistics problem might be in dispersy, I will check that later as well.

**Comment by LipuFei** [comment]
@NielsZeilemaker Ah I see, the problem is this line:
```
SearchGridManager:1739  ChannelManager: searchLocalDB: returning old hit list 0
```
Because `searchkeywords` is the same as `oldsearchkeywords`, it simply ignored the database search request.

We either remove this optimization, or we add one more condition: `hits > 0`, so we must already have some hits if we want to ignore the database search request.

**Comment by NielsZeilemaker** [comment]
That shouldn't be a problem, as they are remotehits which are joined with the local results in memory.
Starting at https://github.com/Tribler/tribler/blob/devel/Tribler/Main/vwxGUI/SearchGridManager.py#L1670

**Comment by LipuFei** [comment]
@NielsZeilemaker I've running the test for many times, and I found that, if I search only once, it will probably fail even if we do get some channel results, but if I manually search again, the Channel item will appear on the GUI. Do you know what part of the code is doing the GUI list update? I tried `refresh()` in list.py but it didn't work.

**Pull request closed without merging**